File: deprecated/distnetcache/dynamic_array.c
#include "../common/dynamic_array.h"

DynamicArray::DynamicArray() {
	_cursize = 0;
	_curcapability = 0;
	_array = NULL;

	_mincapability = 0;
	_maxcapability = 0;
}

DynamicArray::DynamicArray(int mincapability, int maxcapability) {
	_cursize = 0;
	init(mincapability, maxcapability);
}

DynamicArray::~DynamicArray() {
	if (_array != NULL) {
		delete [] _array;
		_array = NULL;
	}
}

void DynamicArray::init(int mincapability, int maxcapability) {
	INVARIANT(maxcapability > mincapability);

	_mincapability = mincapability;
	_maxcapability = maxcapability;

	_curcapability = mincapability;
	_array = new char[_curcapability];
	memset(_array, 0, _curcapability);
}

char& DynamicArray::operator[](int idx) {
	INVARIANT(idx >= 0 && idx < _cursize);
	return _array[idx];
}

void DynamicArray::dynamic_memcpy(int off, char *srcarray, int len) {
	dynamic_reserve(off, len);

	memcpy(_array + off, srcarray, len);
	if (off + len > _cursize) {
		_cursize = off + len;
	}
}

void DynamicArray::dynamic_memset(int off, int value, int len) {
	dynamic_reserve(off, len);

	memset(_array + off, value, len);
	if (off + len > _cursize) {
		_cursize = off + len;
	}
}

void DynamicArray::clear() {
	if (_curcapability != _mincapability) {
		INVARIANT(_array != NULL);
		delete [] _array;
		_array = new char[_mincapability];
		_curcapability = _mincapability;
	}
	memset(_array, 0, _mincapability);
	_cursize = 0;
}

int DynamicArray::size() const {
	return _cursize;
}

char *DynamicArray::array() const {
	return _array;
}

void DynamicArray::dynamic_reserve(int off, int len) {
	if (off + len > _maxcapability) {
		printf("[DynamicArray] off %d + len %d exceeds maxcapability %d!\n", off, len, _maxcapability);
		exit(-1);
	}

	// [off, off+len)
	if (off + len > _curcapability) {
		// NOTE: _curcapability < off + len <= _maxcapability
		INVARIANT(_curcapability < _maxcapability);

		int newcapability = 2*_curcapability;
		while (newcapability < off + len) {
			newcapability = 2*newcapability;
		}
		if (newcapability > _maxcapability) {
			newcapability = _maxcapability;
		}

		char *newarray = new char[newcapability];
		memset(newarray, 0, newcapability);
		memcpy(newarray, _array, _curcapability);
		delete [] _array;
		_array = newarray;
		newarray = NULL;
		_curcapability = newcapability;
	}
}

File: benchmark/inswitchcache-java-lib/src/main/java/com/inswitchcache/core/DynamicArray.java
package com.inswitchcache.core;

import java.util.Arrays;

/**
 * Class for sending/receiving request/response with large value.
 */
public final class DynamicArray {
  private int cursize = 0;
  private int curcapability = 0;
  private int mincapability = 0;
  private int maxcapability = 0;
  private byte[] array = null;

  public DynamicArray() {
    cursize = 0;
    curcapability = 0;
    mincapability = 0;
    maxcapability = 0;
    array = null;
  }

  public DynamicArray(int tmpmincapability, int tmpmaxcapability) {
    cursize = 0;
    init(tmpmincapability, tmpmaxcapability);
  }

  @Override
  protected void finalize() {
    array = null; // free memory of array
  }

  public void init(int tmpmincapability, int tmpmaxcapability) {
    MyUtils.myAssert(tmpmaxcapability > tmpmincapability);

    mincapability = tmpmincapability;
    maxcapability = tmpmaxcapability;

    curcapability = mincapability;
    array = new byte[mincapability];
    Arrays.fill(array, (byte) 0);
  }

  public byte get(int idx) {
    MyUtils.myAssert(idx >= 0 && idx < cursize);
    return array[idx];
  }

  public void set(int idx, byte val) {
    MyUtils.myAssert(idx >= 0 && idx < cursize);
    array[idx] = val;
  }

  public void dynamicMemcpy(int off, byte[] srcarray, int len) {
    dynamicReserve(off, len);

    System.arraycopy(srcarray, 0, array, off, len);
    if (off + len > cursize) {
      cursize = off + len;
    }
  }

  public void dynamicMemset(int off, int value, int len) {
    dynamicReserve(off, len);

    Arrays.fill(array, off, off + len, (byte) value);
    if (off + len > cursize) {
      cursize = off + len;
    }
  }

  public void clear() {
    if (curcapability != mincapability) {
      MyUtils.myAssert(array != null);
      array = null; // free memory of array
      array = new byte[mincapability];
      curcapability = mincapability;
    }
    Arrays.fill(array, (byte) 0);
    cursize = 0;
  }

  public int size() {
    return cursize;
  }

  public byte[] array() {
    return array;
  }

  public void dynamicReserve(int off, int len) {
    if (off + len > maxcapability) {
      System.out
          .println(String.format("[DynamicArray] off %d + len %d exceeds maxcapability %d", off, len, maxcapability));
      System.out.flush();
      System.exit(-1);
    }

    if (off + len > curcapability) {
      // NOTE: curcapability < off + len <= maxcapability
      MyUtils.myAssert(curcapability < maxcapability);

      int newcapability = 2 * curcapability;
      while (newcapability < off + len) {
        newcapability = 2 * newcapability;
      }
      if (newcapability > maxcapability) {
        newcapability = maxcapability;
      }

      byte[] newarray = new byte[newcapability];
      Arrays.fill(newarray, (byte) 0);
      System.arraycopy(array, 0, newarray, 0, curcapability);
      array = null; // free memory of array
      array = newarray; // shallow copy
      newarray = null; // NOT free memory of newarray
      curcapability = newcapability;
    }
  }
}
